gufe/storage/pseudodirectory.py

```diff
@@ -0,0 +1,156 @@
+        self.shared_dir = self.scratch / holding / prefix
+        self.shared_dir.mkdir(exist_ok=True, parents=True)
+
+    def get_other_shared_dir(self, prefix, delete_holding=None):
+        """Get a related unit's shared directory.
+        """
+        if delete_holding is None:
+            delete_holding = self.delete_holding
+
+        return SharedRoot(
+            scratch=self.scratch,
+            external=self.external,
+            prefix=prefix,
+            holding=self.holding,
+            delete_holding=delete_holding,
+            read_only=True,
+        )
+
+    def transfer_holding_to_external(self):
+        """Transfer all objects in the registry to external storage"""
+        if self.read_only:
+            logging.debug("Read-only: Not transfering to external storage")
+            return  # early exit
+
+        for obj in self.registry:
+            path = Path(obj)
+            if not path.exists():
+                logging.info(f"Found nonexistent path {path}, not "
+                             "transfering to external storage")
+            elif path.is_dir():
+                logging.debug(f"Found directory {path}, not "
+                             "transfering to external storage")
+            else:
+                logging.info(f"Transfering {path} to external storage")
+                self.external.store_path(obj.label, path)
+
+    def __del__(self):
+        # take everything in self.shared_dir and write to it shared; keeping
+        # our prefix
+        self.transfer_holding_to_external()
+        if self.delete_holding:
+            shutil.rmtree(self.shared_dir)
+
+    def register_path(self, shared_path):
+        label_exists = self.external.exists(shared_path.label)
+
+        if self.read_only and not label_exists:
+            raise IOError(f"Unable to create '{shared_path.label}'. This "
+                          "shared path is read-only.")
+
+        self.registry.add(shared_path)
+
+        # if this is a file that exists, bring it into our subdir
+        # NB: this happens even if you're intending to overwrite the path,
+        # which is kind of wasteful
+        if label_exists:
+            scratch_path = self.shared_dir / shared_path.path
+            # TODO: switch this to using `get_filename` and `store_path`
+            with self.external.load_stream(shared_path.label) as f:
+                external_bytes = f.read()
+            if scratch_path.exists():
+                ... # TODO: something to check that the bytes are the same?
+            scratch_path.parent.mkdir(exist_ok=True, parents=True)
+            with open(scratch_path, mode='wb') as f:
+                f.write(external_bytes)
+
+    def __truediv__(self, path: PathLike):
+        return SharedPath(root=self, path=path)
+
+    def __fspath__(self):
+        return str(self.shared_dir)
+
+    def __repr__(self):
+        return f"SharedRoot({self.scratch}, {self.external}, {self.prefix})"
+
+
+class SharedPath:
```

Comment:
why is this necessary?  Could `SharedRoot` just hand out real pathlikes?  Is this class just to avoid literally giving out Paths?

Reply:
```python
filepath = root / "directory" / "file.txt"
```

if root/directory is a `pathlib.Path`, then root/directory is registered, but root/directory/file.txt is not. So we need the purpose is to use our `__truediv__` operator and register with the root.